controllers/csi/daemonset.go

```diff
@@ -0,0 +1,356 @@
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type Reconciler struct {
+	client   client.Client
+	scheme   *runtime.Scheme
+	logger   logr.Logger
+	instance *v1alpha1.DynaKube
+}
+
+func NewReconciler(client client.Client, scheme *runtime.Scheme, logger logr.Logger, instance *v1alpha1.DynaKube) *Reconciler {
+	return &Reconciler{
+		client:   client,
+		scheme:   scheme,
+		logger:   logger,
+		instance: instance,
+	}
+}
+
+func (r *Reconciler) Reconcile() (bool, error) {
+	r.logger.Info("Reconciling CSI driver")
+
+	operatorImage, err := r.getOperatorImage()
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	ds, err := buildDesiredCSIDaemonSet(operatorImage)
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	if err := controllerutil.SetControllerReference(r.instance, ds, r.scheme); err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	upd, err := utils.CreateOrUpdateDaemonSet(r.client, r.logger, ds)
+	if upd || err != nil {
+		return upd, errors.WithStack(err)
+	}
+
+	return false, nil
+}
+
+func (r *Reconciler) getOperatorImage() (string, error) {
+	var operatorPod v1.Pod
+	operatorPodName := os.Getenv("POD_NAME")
+	operatorNamespace := os.Getenv("POD_NAMESPACE")
+
+	if err := r.client.Get(context.TODO(), client.ObjectKey{Name: operatorPodName, Namespace: operatorNamespace}, &operatorPod); err != nil {
+		return "", errors.WithStack(err)
+	}
+
+	return operatorPod.Spec.Containers[0].Image, nil
+}
+
+func buildDesiredCSIDaemonSet(operatorImage string) (*appsv1.DaemonSet, error) {
+	metadata := prepareMetadata()
+	driver := prepareDriverSpec(operatorImage)
+	registrar := prepareRegistrarSpec(operatorImage)
+	livenessprobe := preparelivenessProbeSpec(operatorImage)
+	vol := prepareVolumes()
+
+	ds := &appsv1.DaemonSet{
+		ObjectMeta: metadata,
+		Spec: appsv1.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"internal.oneagent.dynatrace.com/component": "csi-driver",
+					"internal.oneagent.dynatrace.com/app":       "csi-driver",
+				},
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						"kubectl.kubernetes.io/default-logs-container": "driver",
+					},
+					Labels: map[string]string{
+						"internal.oneagent.dynatrace.com/component": "csi-driver",
+						"internal.oneagent.dynatrace.com/app":       "csi-driver",
+					},
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						driver,
+						registrar,
+						livenessprobe,
+					},
+					ServiceAccountName: DaemonSetName,
+					Volumes:            vol,
+				},
+			},
+		},
+	}
```

Comment:
creation of DaemonSet could be wrapped in a function, but in that case that's more of a nitpick